secboot/luks2/cryptsetup.go

```diff
@@ -142,77 +121,36 @@ func AddKey(devicePath string, existingKey, key []byte, options *AddKeyOptions)
 		options = &AddKeyOptions{Slot: AnySlot}
 	}
 
-	fifoPath, cleanupFifo, err := mkFifo()
-	if err != nil {
-		return xerrors.Errorf("cannot create FIFO for passing existing key to cryptsetup: %w", err)
-	}
-	defer cleanupFifo()
-
 	args := []string{
 		// add a new key
 		"luksAddKey",
 		// LUKS2 only
 		"--type", "luks2",
-		// read existing key from named pipe
-		"--key-file", fifoPath}
+		// read existing key from stdin, specifying size so no new line
+		// separator is needed (as the existing key can contain them)
+		"--key-file", "-",
+		"--keyfile-size", strconv.Itoa(len(existingKey)),
+		// remove warnings and confirmation questions
+		"--batch-mode"}
 
 	// apply KDF options
 	args = options.KDFOptions.appendArguments(args)
 
 	if options.Slot != AnySlot {
+		// TODO use --new-key-slot for newer crypsetup versions
 		args = append(args, "--key-slot", strconv.Itoa(options.Slot))
 	}
 
 	args = append(args,
 		// container to add key to
 		devicePath,
-		// read new key from stdin.
-		// Note that we can't supply the new key and existing key via the same channel
-		// because pipes and FIFOs aren't seekable - we would need to use an actual file
-		// in order to be able to do this.
-		"-")
-
-	cmd := exec.Command("cryptsetup", args...)
-	cmd.Stdin = bytes.NewReader(key)
-
-	// Writing to the fifo must happen in a go-routine as it may block
-	// if the other side is not connected. Special care must be taken
-	// about the cleanup.
-	fifoErrCh := make(chan error)
-	go func() {
-		fifoErr := writeExistingKeyToFifo(fifoPath, existingKey)
-		if fifoErr != nil {
-			// kill to ensure cmd is not lingering
-			if cmd.Process != nil {
-				cmd.Process.Kill()
-			}
-			// also ensure fifo is closed
-			cleanupFifo()
-		}
-		fifoErrCh <- fifoErr
-	}()
-
-	output, cmdErr := cmd.CombinedOutput()
-	if cmdErr != nil {
-		// cleanupFifo will open/close the fifo to ensure the
-		// writeExistingKeyToFifo() does not leak while waiting
-		// for the other side of the fifo to connect (it may never
-		// do)
-		cleanupFifo()
-	}
-	fifoErr := <-fifoErrCh
+		// we read raw bytes up to EOF (so new key can contain '\n')
+		"-",
```

Comment:
Are we providing the unlock key and the new key both on the stdin?

Is that why then we need `--keyfile-size`?